content/blog/2025-03-27-valkey-8-1-0-rc1.md

```diff
@@ -0,0 +1,282 @@
+over Valkey 8.0 for all Valkey installations. 
+In this blog, we'll dive a bit deeper into some of these areas 
+and talk about the exciting features built for this release.
+
+## Performance
+
+There are many changes in 8.1 that improve Valkey's performance either by decreasing the latency of some operations, increasing the throughput of some operations and reducing the memory footprint of the overall system.
+
+
+### The New Hashtable
+
+One of the main changes responsible for several performance improvements is the new hashtable implementation that is used as the main key-value store in Valkey and also used in the Hash, Set, and Sorted Set data types implementation.
+
+Valkey is a distributed key-value and therefore is no surprise that a hashtable data structure performs an important role in the way Valkey works.
+
+Almost all commands go through the main key-value store where the several datatypes values are stored, and therefore the hashtable implementation is present in the critical path of most workloads.
+
+The new hashtable implementation is a complete rewrite of the previous hashtable implementation, which reduces the memory usage by roughly 20 bytes per key-value pair. The new implementation also helps improve the Valkey server supported throughput by roughly 10% compared to 8.0 version for pipeline workloads when I/O threading is not used.
+
+The previous hashtable implementation, called `dict`, had the traditional design of an array of buckets where each bucket is a collision list. Each collision list element is called a `dictEntry` and holds a single key-value pair. The `dictEntry` structure stores three pointers, one for the key, one for the value, and one for the next `dictEntry` in the collision list. In terms of memory consumption, in a 64bit architecture, each `dictEntry` takes 24 bytes of memory.
+
+![Dictionary structure](/assets/media/pictures/dict_structures.png)
+
+Regarding memory accesses, the above design requires at least 4 memory accesses to get the value in the case that the key is stored in the first `dictEntry` of the collision list. This might have a significant impact in cache locality when traversing the collision list search for the right key.
+
+The new hashtable implementation improves on the memory footprint required and on the cache locality by taking a different approach on the design of each bucket. Each bucket has a fixed size of 64 bytes (in a 64bit architecture), which usually corresponds to the size of a single cache line. The bucket structure stores 7 pointers, 8 bytes each, plus 8 bytes used for metadata. The 7 pointers are used to store up to 7 entries. Keys that map to the same bucket are all stored in the same bucket. If more than 7 keys map to the same bucket, a new bucket is linked to the previous one using the last of the 7 pointers. Each entry pointer, points to an object that may embed the key and the value.
```

Comment:
there is quite some detail in these paragraphs. @zuiderkwast, how do you like to cover this part to reduce redundancy? I was thinking that we could focus more on the results/impacts and the high level ideas in this blog and leave the details like data structures etc to yours. this would also align better with other sections such as "Iterator Prefetching"

Reply:
Yeah I think we should remove most of the text about the internals of the hash table. Only the key takeaway are needed in this release blog post and a link to the hash table dedicated blog post.